internal/pkg/agent/application/upgrade/step_download.go

```diff
@@ -56,7 +62,7 @@ func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI stri
 		return "", errors.New(err, fmt.Sprintf("failed to create download directory at %s", paths.Downloads()))
 	}
 
-	path, err := fetcher.Download(ctx, agentArtifact, version)
+	path, err := fetcher.Download(ctx, agentArtifact, parsedVersion.VersionWithPrerelease())
```

Comment:
The reason why we need the `VersionWithPrerelease()` is that, regardless of the buildId our artifact are always named `elastic-agent-x.y.z[-SNAPSHOT]` without build meta information.
The method is there to give out just that version format omitting the build metadata that may be part of the version string
Before the change the `version` string would be enough as we never specified a build id.

I will check the omission in the `newVerifier` as it should follow the same pattern

Reply:
Checked the newVerifier call: the snapshot.Verifier object uses the version only to determine the snapshotURI (same function as called by the snapshot.Downloader) where we need the core version and the build metadata to determine the snapshot URI. A few lines below (line 79 at the time of this comment) we pass the same `parsedVersion.VersionWithPrerelease()` in the `Verify()` method call

Reply:
We need to document these expectations in the code. 

My first suggestion would be to just pass a `ParsedSemVer` type through without calling any methods, and let the implementation of each function use the methods on that type to extract what it needs. Right now you have to read the implementations of each of the functions to know if it is correct, it seems simpler to just push this method call into the implementation itself.

Reply:
Went with a simple comment about the reason why we pass the `VersionWithPrerelease()` version string.
Passing the parsed version wouldn't work because of the way the various downloaders use each other's implementation (the snapshot and the base downloaders would have to implement special behaviour to distinguish between SNAPSHOT or not and that change propagates fast to all others.